File: problems/python/src/longest_substring_without_repeating_characters.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def length_of_longest_substring(self, s: str) -> int:
        pass


class Default(Problem):
    @override
    def length_of_longest_substring(self, s: str) -> int:
        if not s:
            return 0
        # unique characters
        chars = set()
        start = end = 0
        max_length = float('-inf')
        while end < len(s):
            if s[end] not in chars:
                chars.add(s[end])
                end += 1
                max_length = max(max_length, len(chars))
                continue
            chars.remove(s[start])
            start += 1
        return max_length


SOLUTIONS: list[Problem] = [Default()]

File: java/src/main/java/problems1/LongestSubstringWithoutRepeatingCharacters.java
package problems1;

import java.util.HashSet;
import java.util.Set;

public enum LongestSubstringWithoutRepeatingCharacters {
    DEFAULT {
        @Override
        public int lengthOfLongestSubstring(String s) {
            if (s.isEmpty()) {
                return 0;
            }
            // unique characters
            Set<Character> characters = new HashSet<>();

            int start = 0;
            int end = 0;
            int maxLength = Integer.MIN_VALUE;
            while (end < s.length()) {
                if (characters.add(s.charAt(end))) {
                    maxLength = Math.max(maxLength, characters.size());
                    end++;
                    continue;
                }
                characters.remove(s.charAt(start));
                start++;
            }
            return maxLength;
        }
    };

    public abstract int lengthOfLongestSubstring(String s);
}
